File: leetcode/isIsomorphic/isIsomorphic.py
# Given two strings s and t, determine if they are isomorphic.

# Two strings are isomorphic if the characters in s can be replaced to get t.

# All occurrences of a character must be replaced with another character 
# while preserving the order of characters. 
# No two characters may map to the same character but a character may map to itself.

# Example 1:
# Input: s = "egg", t = "add"
# Output: true

# Example 2:
# Input: s = "foo", t = "bar"
# Output: false

# Example 3:
# Input: s = "paper", t = "title"
# Output: true
# Note:
# You may assume both s and t have the same length.


class Solution:
    
    def isIsomorphic(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        t_unique = set(t)
        for c in t_unique:
            idx_t = [i for i in range(len(s)) if t[i] == c]
            c_s = s[idx_t[0]]
            idx_s = [i for i in range(len(t)) if s[i] == c_s]
            if len(idx_t) != len(idx_s) or idx_t != idx_s:
                return False
        return True

if __name__ == '__main__':
    solver = Solution()
    print(solver.isIsomorphic(s = "egg", t = "add"))
    print(solver.isIsomorphic(s = "foo", t = "bar"))
    print(solver.isIsomorphic(s = "paper", t = "title"))


# https://leetcode.com/problems/isomorphic-strings/discuss/57941/Python-different-solutions-(dictionary-etc).
def isIsomorphic1(self, s, t):
    d1, d2 = {}, {}
    for i, val in enumerate(s):
        d1[val] = d1.get(val, []) + [i]
    for i, val in enumerate(t):
        d2[val] = d2.get(val, []) + [i]
    return sorted(d1.values()) == sorted(d2.values())

def isIsomorphic4(self, s, t): 
    return [s.find(i) for i in s] == [t.find(j) for j in t]

            

File: leetcode/isIsomorphic/isIsomorphic.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
using namespace std;

// !!! Time Limit Exceeded
// class Solution {
// public:
//     bool isIsomorphic(string s, string t) {
//         vector<char> t_unique;
//         unique_copy(t.begin(), t.end(), back_inserter(t_unique));
//         for(auto c: t_unique){
//             vector<int> idx_t;
//             for(int i=0; i< int(t.size()); ++i){
//                 if(c == t[i])
//                     idx_t.push_back(i);
//             }

//             vector<int> idx_s;
//             char c_s = s[idx_t[0]];
//             for(int i=0; i< int(s.size()); ++i){
//                 if(c_s == s[i])
//                     idx_s.push_back(i);
//             }
//             if(idx_s.size() != idx_t.size() || idx_s != idx_t)
//                 return false;
//         }
//         return true;    
//     }
// };


// There are 256 ASCII characters. 
// This includes standard ASCII characters(0-127) and Extended ASCII characters(128-255)
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        map<char, vector<int>> d1;
        map<char, vector<int>> d2;
        for(int i=0; i<int(s.size()); ++i){
                d1[s[i]].push_back(i);
                d2[t[i]].push_back(i);   
        }
        vector<vector<int>> d1_vals;
        vector<vector<int>> d2_vals;
        for (auto const& element : d1) {
            d1_vals.push_back(element.second);
        }
        for (auto const& element : d2) {
            d2_vals.push_back(element.second);
        }
        sort(d1_vals.begin(), d1_vals.end());
        sort(d2_vals.begin(), d2_vals.end());
        return d1_vals == d2_vals;

    }
};


int main(){
    // string t("add");
    // vector<char> t_unique;
    // unique_copy(t.begin(), t.end(), back_inserter(t_unique));
    // cout << t << endl;
    // for(auto x: t_unique)
    //     cout << x << endl;

    Solution solver;
    cout << solver.isIsomorphic("egg", "add") << endl;
    cout << solver.isIsomorphic("foo", "bar") << endl;
    cout << solver.isIsomorphic("paper", "title") << endl;
    // map<char, vector<int>> d1;
    // d1['a'].push_back(1);
    // d1['a'].push_back(2);
    // cout << d1['a'][0] << endl;
    // cout << d1['a'].size() << endl;


}